internal/builtin/providers/terraform/resource_data.go

```diff
@@ -74,8 +74,10 @@ func planDataStoreResourceChange(req providers.PlanResourceChangeRequest) (resp
 		// Set the id value to unknown.
 		planned["id"] = cty.UnknownVal(cty.String)
 
-		// Only compute a new output if input has a non-null value.
-		if !input.IsNull() {
+		// Output type must always match the input, even when it's null.
+		if input.IsNull() {
+			planned["output"] = cty.NullVal(input.Type())
```

Comment:
If we already know the input is null from the check above, isn't this equivalent to `planned["output"] = input`?

(If it isn't equivalent for an edge-case reason I'm not considering then I think this version is still better because it will preserve the given value exactly, and so better comply with the spirit of the lifecycle rules)

Reply:
lol, yes that is better. I was only copied from below, where I missed it the first time ;)